File: src/services/savedCommand.ts
/*
 * Kerminal - Modern Terminal Emulator & SSH Manager
 * Copyright (C) 2026 Bùi Thanh Xuân (klpod221)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { api } from "./api";
import type {
  SavedCommand,
  SavedCommandGroup,
  CreateSavedCommandRequest,
  UpdateSavedCommandRequest,
  CreateSavedCommandGroupRequest,
  UpdateSavedCommandGroupRequest,
} from "../types/savedCommand";

/**
 * Saved Command service for frontend
 */
export const savedCommandService = {
  /**
   * Create new saved command
   */
  async createCommand(
    request: CreateSavedCommandRequest,
  ): Promise<SavedCommand> {
    return await api.call("create_saved_command", request);
  },

  /**
   * Get all saved commands
   */
  async getCommands(): Promise<SavedCommand[]> {
    return await api.callRaw("get_saved_commands");
  },

  /**
   * Get saved command by ID
   */
  async getCommand(id: string): Promise<SavedCommand> {
    return await api.callRaw("get_saved_command", id);
  },

  /**
   * Update saved command
   */
  async updateCommand(
    id: string,
    request: UpdateSavedCommandRequest,
  ): Promise<SavedCommand> {
    return await api.callRaw("update_saved_command", id, request);
  },

  /**
   * Delete saved command
   */
  async deleteCommand(id: string): Promise<void> {
    return await api.callRaw("delete_saved_command", id);
  },

  /**
   * Increment command usage count
   */
  async incrementUsage(id: string): Promise<void> {
    return await api.callRaw("increment_command_usage", id);
  },

  /**
   * Toggle command favorite status
   */
  async toggleFavorite(id: string): Promise<SavedCommand> {
    return await api.callRaw("toggle_command_favorite", id);
  },

  /**
   * Create new saved command group
   */
  async createGroup(
    request: CreateSavedCommandGroupRequest,
  ): Promise<SavedCommandGroup> {
    return await api.call("create_saved_command_group", request);
  },

  /**
   * Get all saved command groups
   */
  async getGroups(): Promise<SavedCommandGroup[]> {
    return await api.callRaw("get_saved_command_groups");
  },

  /**
   * Get saved command group by ID
   */
  async getGroup(id: string): Promise<SavedCommandGroup> {
    return await api.callRaw("get_saved_command_group", id);
  },

  /**
   * Update saved command group
   */
  async updateGroup(
    id: string,
    request: UpdateSavedCommandGroupRequest,
  ): Promise<SavedCommandGroup> {
    return await api.callRaw("update_saved_command_group", id, request);
  },

  /**
   * Delete saved command group
   */
  async deleteGroup(id: string): Promise<void> {
    return await api.callRaw("delete_saved_command_group", id);
  },
};

File: src-tauri/src/commands/database/saved_command.rs
use crate::models::saved_command::{
    CreateSavedCommandGroupRequest, CreateSavedCommandRequest, SavedCommand, SavedCommandGroup,
    UpdateSavedCommandGroupRequest, UpdateSavedCommandRequest,
};
use crate::state::AppState;
use tauri::State;

use crate::commands::database::common::app_result;

/// Macro for saved command service results that return String errors
macro_rules! saved_command_result {
    ($expr:expr) => {
        $expr
    };
}

// === Saved Command Commands ===

/// Create new saved command
#[tauri::command]
pub async fn create_saved_command(
    state: State<'_, AppState>,
    request: CreateSavedCommandRequest,
) -> Result<SavedCommand, String> {
    app_result!(state.saved_command_service.create_command(request).await)
}

/// Get all saved commands
#[tauri::command]
pub async fn get_saved_commands(state: State<'_, AppState>) -> Result<Vec<SavedCommand>, String> {
    app_result!(state.saved_command_service.get_commands().await)
}

/// Update saved command
#[tauri::command]
pub async fn update_saved_command(
    state: State<'_, AppState>,
    id: String,
    request: UpdateSavedCommandRequest,
) -> Result<SavedCommand, String> {
    app_result!(
        state
            .saved_command_service
            .update_command(&id, request)
            .await
    )
}

/// Delete saved command
#[tauri::command]
pub async fn delete_saved_command(state: State<'_, AppState>, id: String) -> Result<(), String> {
    app_result!(state.saved_command_service.delete_command(&id).await)
}

/// Increment command usage count
#[tauri::command]
pub async fn increment_command_usage(state: State<'_, AppState>, id: String) -> Result<(), String> {
    app_result!(state.saved_command_service.increment_usage(&id).await)
}

/// Toggle command favorite status
#[tauri::command]
pub async fn toggle_command_favorite(
    state: State<'_, AppState>,
    id: String,
) -> Result<SavedCommand, String> {
    app_result!(state.saved_command_service.toggle_favorite(&id).await)
}

// === Saved Command Group Commands ===

/// Create new saved command group
#[tauri::command]
pub async fn create_saved_command_group(
    state: State<'_, AppState>,
    request: CreateSavedCommandGroupRequest,
) -> Result<SavedCommandGroup, String> {
    app_result!(state.saved_command_service.create_group(request).await)
}

/// Get all saved command groups
#[tauri::command]
pub async fn get_saved_command_groups(
    state: State<'_, AppState>,
) -> Result<Vec<SavedCommandGroup>, String> {
    app_result!(state.saved_command_service.get_groups().await)
}

/// Update saved command group
#[tauri::command]
pub async fn update_saved_command_group(
    state: State<'_, AppState>,
    id: String,
    request: UpdateSavedCommandGroupRequest,
) -> Result<SavedCommandGroup, String> {
    app_result!(state.saved_command_service.update_group(&id, request).await)
}

/// Delete saved command group
#[tauri::command]
pub async fn delete_saved_command_group(
    state: State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    app_result!(state.saved_command_service.delete_group(&id).await)
}
